src/garage/replay_buffer/her_replay_buffer.py

```diff
@@ -2,145 +2,175 @@
-            dict[numpy.ndarray]: Transitions.
+            np.ndarray: A numpy array of HER goals with shape
+                (replay_k, goal_dim).
 
         """
-        # Select which episodes to use
-        time_horizon = episode_batch['action'].shape[1]
-        rollout_batch_size = episode_batch['action'].shape[0]
-        episode_idxs = np.random.randint(rollout_batch_size,
-                                         size=sample_batch_size)
-        # Select time steps to use
-        t_samples = np.random.randint(time_horizon, size=sample_batch_size)
-        transitions = {
-            key: episode_batch[key][episode_idxs, t_samples]
-            for key in episode_batch.keys()
-        }
-
-        her_idxs = np.where(
-            np.random.uniform(size=sample_batch_size) < future_p)
-        future_offset = np.random.uniform(
-            size=sample_batch_size) * (time_horizon - t_samples)
-        future_offset = future_offset.astype(int)
-        future_t = (t_samples + future_offset)[her_idxs]
-
-        future_ag = episode_batch['achieved_goal'][episode_idxs[her_idxs],
-                                                   future_t]
-        transitions['goal'][her_idxs] = future_ag
-
-        achieved_goals = episode_batch['achieved_goal'][episode_idxs[her_idxs],
-                                                        t_samples[her_idxs]]
-        transitions['achieved_goal'][her_idxs] = achieved_goals
-
-        # Re-compute reward since we may have substituted the goal.
-        reward_params_keys = inspect.signature(reward_fun).parameters.keys()
-        reward_params = {
-            rk: transitions[k]
-            for k, rk in zip(['next_achieved_goal', 'goal'],
-                             list(reward_params_keys)[:-1])
-        }
-        reward_params['info'] = {}
-        transitions['reward'] = reward_fun(**reward_params)
-
-        transitions = {
-            k: transitions[k].reshape(sample_batch_size,
-                                      *transitions[k].shape[1:])
-            for k in transitions.keys()
-        }
-
-        goals = transitions['goal']
-        next_inputs = np.concatenate((transitions['next_observation'], goals,
-                                      transitions['achieved_goal']),
-                                     axis=-1)
-        inputs = np.concatenate(
-            (transitions['observation'], goals, transitions['achieved_goal']),
-            axis=-1)
-        transitions['observation'] = inputs
-        transitions['next_observation'] = next_inputs
-
-        assert transitions['action'].shape[0] == sample_batch_size
-        return transitions
-
-    return _her_sample_transitions
+        goal_indexes = np.random.randint(transition_idx + 1,
+                                         len(path['observations']),
+                                         size=self._replay_k)
+        return path['achieved_goals'][goal_indexes]
 
+    def sample_transitions(self, batch_size):
```

Comment:
It's a little weird to me to define this method. You could instead just pre-process the transitions you store in `add_path`, instead of re-processing them here every time they're sampled.

I'm happy to merge this as-is, but it doesn't quite seem ideal.

Reply:
True, fixed.